### ISSUE TITLE
When given incomplete variable definition, parser returns incomplete AST instead of throwing 

### ISSUE THREAD

**Issue by DanielSWolf** [typescript-eslint/typescript-eslint#3326]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

This issue is *not* about a `@typescript-eslint` linter rule, but about the parser `@typescript-eslint/parser` itself. Consider the following test code:

```js
const { parseForESLint } = require('@typescript-eslint/parser');

describe('parseForESLint', () => {
  it('returns AST for valid code', () => {
    expect(() => parseForESLint('let foo;')).not.toThrow();
  });

  describe('throws for invalid code', () => {
    it.each([
      '}',        // invalid syntax
      'function', // incomplete function definition
      'let',      // incomplete variable definition
      'var',      // incomplete variable definition
      'const'     // incomplete constant definition
    ])('%p', input => {
      expect(() => parseForESLint(input)).toThrow();
    })
  })
});
```

**Expected Result**

I expected all tests to succeed. All arguments to the parameterized test represent invalid code, so `parseForESLint()` should throw in all these cases.

**Actual Result**

As expected, `parseForESLint()` throws for inputs `'}'` (invalid syntax) and `'function'` (incomplete function). However, it does *not* throw for the incomplete variable/constant definitions `'let'`, `'var'`, and `'const'`. Instead, it returns an incomplete AST.

The problem is that tools further down the line don't expect to receive an incomplete AST. In particular, [here is an issue](https://github.com/eslint/eslint/issues/14538) for the ESLint rule `indent` which throws an exception because `node.declarations` is a zero-length array.

**Additional Info**

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.22.0` |
| `TypeScript`                | `4.2.4` |
| `ESLint`                    | n/a |
| `node`                      | `14.16.0` |

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/1852
